File: packages/capacitor-notificare-geo/ios/Plugin/NotificareGeoPlugin.swift
import Foundation
import Capacitor
import CoreLocation
import NotificareKit
import NotificareGeoKit

private let REQUESTED_LOCATION_ALWAYS_KEY = "re.notifica.geo.capacitor.requested_location_always"

@objc(NotificareGeoPlugin)
public class NotificareGeoPlugin: CAPPlugin {
    private let locationManager = CLLocationManager()
    private var requestedPermission: PermissionGroup?
    private var requestedPermissionCall: CAPPluginCall?
    
    private var authorizationStatus: CLAuthorizationStatus {
        if #available(iOS 14.0, *) {
            return locationManager.authorizationStatus
        }
        
        return CLLocationManager.authorizationStatus()
    }
    
    public override func load() {
        EventBroker.instance.setup { self.notifyListeners($0, data: $1) }
        Notificare.shared.geo().delegate = self
        
        locationManager.delegate = self
    }
    
    @objc func hasLocationServicesEnabled(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.geo().hasLocationServicesEnabled
        ])
    }
    
    @objc func hasBluetoothEnabled(_ call: CAPPluginCall) {
        call.resolve([
            "result": Notificare.shared.geo().hasBluetoothEnabled
        ])
    }
    
    @objc func enableLocationUpdates(_ call: CAPPluginCall) {
        Notificare.shared.geo().enableLocationUpdates()
        call.resolve()
    }
    
    @objc func disableLocationUpdates(_ call: CAPPluginCall) {
        Notificare.shared.geo().disableLocationUpdates()
        call.resolve()
    }
    
    @objc func checkPermissionStatus(_ call: CAPPluginCall) {
        guard let permissionStr = call.getString("permission"), let permission = PermissionGroup(rawValue: permissionStr) else {
            call.reject("Missing 'permission' parameter.")
            return
        }
        
        let status = checkPermissionStatus(permission)
        call.resolve(["result": status.rawValue])
    }
    
    @objc func shouldShowPermissionRationale(_ call: CAPPluginCall) {
        call.resolve(["result": false])
    }
    
    @objc func presentPermissionRationale(_ call: CAPPluginCall) {
        call.reject("This method is not implemented in iOS.")
    }
    
    @objc func requestPermission(_ call: CAPPluginCall) {
        guard let permissionStr = call.getString("permission"), let permission = PermissionGroup(rawValue: permissionStr) else {
            call.reject("Missing 'permission' parameter.")
            return
        }
        
        let status = checkPermissionStatus(permission)
        
        if status == .granted || status == .permanentlyDenied || status == .restricted {
            call.resolve(["result": status.rawValue])
            return
        }
        
        if (permission == .locationWhenInUse) {
            if ((Bundle.main.infoDictionary?["NSLocationWhenInUseUsageDescription"]) == nil) {
                call.reject("Missing 'NSLocationWhenInUseUsageDescription' in the app bundle's Info.plist file.")
                return
            }
        }
        
        if (permission == .locationAlways) {
            if ((Bundle.main.infoDictionary?["NSLocationAlwaysAndWhenInUseUsageDescription"]) == nil) {
                call.reject("Missing 'NSLocationAlwaysAndWhenInUseUsageDescription' in the app bundle's Info.plist file.")
                return
            }
            
            if (authorizationStatus == .notDetermined) {
                call.reject("Location 'When in Use' should be requested before 'Location Always' request.")
                return
            }
        }
        
        requestedPermission = permission
        requestedPermissionCall = call
        
        if permission == .locationWhenInUse {
            locationManager.requestWhenInUseAuthorization()
        } else if permission == .locationAlways {
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(applicationDidBecomeActive),
                name: UIApplication.didBecomeActiveNotification,
                object: nil
            )
            
            locationManager.requestAlwaysAuthorization()
            UserDefaults.standard.set(true, forKey: REQUESTED_LOCATION_ALWAYS_KEY)
        }
    }
    
    @objc func openAppSettings(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let url = URL(string: UIApplication.openSettingsURLString), UIApplication.shared.canOpenURL(url) else {
                call.reject("Unable to open the application settings.")
                return
            }
            
            UIApplication.shared.open(url) { success in
                if success {
                    call.resolve()
                } else {
                    call.reject("Unable to open the application settings.")
                }
            }
        }
    }
    
    @objc private func applicationDidBecomeActive() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.didBecomeActiveNotification,
            object: nil
        )
        
        guard let requestedPermissionCall = requestedPermissionCall else {
            return
        }
        
        if (authorizationStatus != .authorizedAlways) {
            requestedPermissionCall.resolve(["result": PermissionStatus.denied.rawValue])
            
            self.requestedPermission = nil
            self.requestedPermissionCall = nil
        }
    }
    
    private func checkPermissionStatus(_ permission: PermissionGroup) -> PermissionStatus {
        guard permission != .bluetoothScan else {
            return .granted
        }
        
        return determinePermissionStatus(permission, authorizationStatus: authorizationStatus)
    }
    
    private func determinePermissionStatus(_ permission: PermissionGroup, authorizationStatus: CLAuthorizationStatus) -> PermissionStatus {
        if permission == .locationAlways {
            switch authorizationStatus {
            case .notDetermined:
                return .denied
            case .restricted:
                return .restricted
            case .denied:
                return .permanentlyDenied
            case .authorizedWhenInUse:
                return UserDefaults.standard.bool(forKey: REQUESTED_LOCATION_ALWAYS_KEY) ? .permanentlyDenied : .denied
            case .authorizedAlways:
                return .granted
            @unknown default:
                return .denied
            }
        }
        
        switch authorizationStatus {
        case .notDetermined:
            return .denied
        case .restricted:
            return .restricted
        case .denied:
            return .permanentlyDenied
        case .authorizedWhenInUse, .authorizedAlways:
            return .granted
        @unknown default:
            return .denied
        }
    }
}

extension NotificareGeoPlugin: NotificareGeoDelegate {
    public func notificare(_ notificareGeo: NotificareGeo, didUpdateLocations locations: [NotificareLocation]) {
        guard let location = locations.first else { return }
        
        do {
            EventBroker.instance.dispatchEvent("location_updated", data: try location.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the location_updated event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didEnter region: NotificareRegion) {
        do {
            EventBroker.instance.dispatchEvent("region_entered", data: try region.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the region_entered event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didExit region: NotificareRegion) {
        do {
            EventBroker.instance.dispatchEvent("region_exited", data: try region.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the region_exited event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didEnter beacon: NotificareBeacon) {
        do {
            EventBroker.instance.dispatchEvent("beacon_entered", data: try beacon.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the beacon_entered event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didExit beacon: NotificareBeacon) {
        do {
            EventBroker.instance.dispatchEvent("beacon_exited", data: try beacon.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the beacon_exited event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didRange beacons: [NotificareBeacon], in region: NotificareRegion) {
        do {
            let data: [String: Any] = [
                "region": try region.toJson(),
                "beacons": try beacons.map { try $0.toJson() },
            ]
            
            EventBroker.instance.dispatchEvent("beacons_ranged", data: data)
        } catch {
            NotificareLogger.error("Failed to emit the beacons_ranged event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didVisit visit: NotificareVisit) {
        do {
            EventBroker.instance.dispatchEvent("visit", data: try visit.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the visit event.", error: error)
        }
    }
    
    public func notificare(_ notificareGeo: NotificareGeo, didUpdateHeading heading: NotificareHeading) {
        do {
            EventBroker.instance.dispatchEvent("heading_updated", data: try heading.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the heading_updated event.", error: error)
        }
    }
}

extension NotificareGeoPlugin: CLLocationManagerDelegate {
    public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        handleAuthorizationChange(status)
    }
    
    @available(iOS 14.0, *)
    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        handleAuthorizationChange(manager.authorizationStatus)
    }
    
    private func handleAuthorizationChange(_ authorizationStatus: CLAuthorizationStatus) {
        if authorizationStatus == .notDetermined {
            // When the user changes to "Ask Next Time" via the Settings app.
            UserDefaults.standard.removeObject(forKey: REQUESTED_LOCATION_ALWAYS_KEY)
        }
        
        guard let requestedPermission = requestedPermission, let requestedPermissionCall = requestedPermissionCall else {
            return
        }
        
        let status = determinePermissionStatus(requestedPermission, authorizationStatus: authorizationStatus)
        
        // If status is 'Permanently Denied' we will consider that result as 'Denied' because the permission just changed
        // This will only affect 'When in Use' permission
        
        if (status == .permanentlyDenied) {
            requestedPermissionCall.resolve(["result": PermissionStatus.denied.rawValue])
        } else {
            requestedPermissionCall.resolve(["result": status.rawValue])
        }
        
        
        self.requestedPermission = nil
        self.requestedPermissionCall = nil
    }
}

extension NotificareGeoPlugin {
    internal enum PermissionGroup: String, CaseIterable {
        case locationWhenInUse = "location_when_in_use"
        case locationAlways = "location_always"
        case bluetoothScan = "bluetooth_scan"
    }
    
    internal enum PermissionStatus: String, CaseIterable {
        case denied = "denied"
        case granted = "granted"
        case restricted = "restricted"
        case permanentlyDenied = "permanently_denied"
    }
}

File: packages/capacitor-notificare-geo/android/src/main/java/re/notifica/geo/capacitor/NotificareGeoPlugin.kt
package re.notifica.geo.capacitor

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.util.Log
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import re.notifica.Notificare
import re.notifica.geo.NotificareGeo
import re.notifica.geo.ktx.geo
import re.notifica.geo.models.NotificareBeacon
import re.notifica.geo.models.NotificareLocation
import re.notifica.geo.models.NotificareRegion
import re.notifica.geo.models.toJson

@CapacitorPlugin(name = "NotificareGeoPlugin")
public class NotificareGeoPlugin : Plugin(), NotificareGeo.Listener {

    private companion object {
        private val TAG = NotificareGeoPlugin::class.java.simpleName
    }

    private var shouldShowRationale = false
    private var hasOnGoingPermissionRequest = false
    private var permissionRequestCall: PluginCall? = null

    private lateinit var permissionsLauncher: ActivityResultLauncher<Array<String>>

    override fun load() {
        logger.hasDebugLoggingEnabled = Notificare.options?.debugLoggingEnabled ?: false

        EventBroker.setup(this::notifyListeners)

        Notificare.geo().removeListener(this)
        Notificare.geo().addListener(this)

        permissionsLauncher = bridge.registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions()
        ) { permissions ->
            val status = permissions
                .all { it.value }
                .let { granted ->
                    if (granted) {
                        PermissionStatus.GRANTED
                    } else {
                        if (!shouldShowRationale &&
                            !ActivityCompat.shouldShowRequestPermissionRationale(activity, permissions.keys.first())
                        ) {
                            PermissionStatus.PERMANENTLY_DENIED
                        } else {
                            PermissionStatus.DENIED
                        }
                    }
                }

            permissionRequestCall?.resolve(JSObject().put("result", status.rawValue))
            permissionRequestCall = null
            hasOnGoingPermissionRequest = false
        }
    }

    @PluginMethod
    public fun hasLocationServicesEnabled(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.geo().hasLocationServicesEnabled)
            }
        )
    }

    @PluginMethod
    public fun hasBluetoothEnabled(call: PluginCall) {
        call.resolve(
            JSObject().apply {
                put("result", Notificare.geo().hasBluetoothEnabled)
            }
        )
    }

    @PluginMethod
    public fun getMonitoredRegions(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", JSArray(Notificare.geo().monitoredRegions.map { it.toJson() }))
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun getEnteredRegions(call: PluginCall) {
        try {
            call.resolve(
                JSObject().apply {
                    put("result", JSArray(Notificare.geo().enteredRegions.map { it.toJson() }))
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun enableLocationUpdates(call: PluginCall) {
        Notificare.geo().enableLocationUpdates()
        call.resolve()
    }

    @PluginMethod
    public fun disableLocationUpdates(call: PluginCall) {
        Notificare.geo().disableLocationUpdates()
        call.resolve()
    }

    @PluginMethod
    public fun checkPermissionStatus(call: PluginCall) {
        val permission = call.getString("permission")
            .let { str ->
                if (str == null) {
                    call.reject("Missing permission parameter.")
                    return
                }

                PermissionGroup.parse(str) ?: run {
                    call.reject("Unsupported permission parameter: $str")
                    return
                }
            }

        val status = determinePermissionStatus(context, permission)
        call.resolve(JSObject().put("result", status.rawValue))
    }

    @PluginMethod
    public fun shouldShowPermissionRationale(call: PluginCall) {
        val permission = call.getString("permission")
            .let { str ->
                if (str == null) {
                    call.reject("Missing permission parameter.")
                    return
                }

                PermissionGroup.parse(str) ?: run {
                    call.reject("Unsupported permission parameter: $str")
                    return
                }
            }

        val activity = activity ?: run {
            logger.warning("Unable to acquire a reference to the current activity.")
            call.reject("Unable to acquire a reference to the current activity.")
            return
        }

        val manifestPermissions = getManifestPermissions(context, permission)

        if (manifestPermissions.isEmpty()) {
            logger.warning("No permissions found in the manifest for $permission")
            call.resolve(JSObject().put("result", false))
            return
        }

        val showRationale = ActivityCompat.shouldShowRequestPermissionRationale(activity, manifestPermissions.first())
        call.resolve(JSObject().put("result", showRationale))
    }

    @PluginMethod
    public fun presentPermissionRationale(call: PluginCall) {
        val permission = call.getString("permission")
            .let { str ->
                if (str == null) {
                    call.reject("Missing permission parameter.")
                    return
                }

                PermissionGroup.parse(str) ?: run {
                    call.reject("Unsupported permission parameter: $str")
                    return
                }
            }

        val rationale = call.getObject("rationale") ?: run {
            call.reject("Missing rationale parameter.")
            return
        }

        val activity = activity ?: run {
            logger.warning("Unable to acquire a reference to the current activity.")
            call.reject("Unable to acquire a reference to the current activity.")
            return
        }

        val title = rationale.getString("title")
        val message = rationale.getString("message")
        val buttonText = rationale.getString("buttonText") ?: context.getString(android.R.string.ok)

        try {
            logger.debug("Presenting permission rationale for '$permission'.")

            activity.runOnUiThread {
                AlertDialog.Builder(activity)
                    .setTitle(title)
                    .setMessage(message)
                    .setCancelable(false)
                    .setPositiveButton(buttonText, null)
                    .setOnDismissListener { call.resolve() }
                    .show()
            }
        } catch (e: Exception) {
            call.reject("Unable to present the rationale alert.", e)
        }
    }

    @PluginMethod
    public fun requestPermission(call: PluginCall) {
        val permission = call.getString("permission")
            .let { str ->
                if (str == null) {
                    call.reject("Missing permission parameter.")
                    return
                }

                PermissionGroup.parse(str) ?: run {
                    call.reject("Unsupported permission parameter: $str")
                    return
                }
            }

        if (hasOnGoingPermissionRequest) {
            logger.warning("A request for permissions is already running, please wait for it to finish before doing another request.")
            call.reject("A request for permissions is already running, please wait for it to finish before doing another request.")
            return
        }

        val status = determinePermissionStatus(context, permission)
        if (status == PermissionStatus.GRANTED) {
            call.resolve(JSObject().put("result", status.rawValue))
            return
        }

        val manifestPermissions = getManifestPermissions(context, permission)

        if (manifestPermissions.isEmpty()) {
            logger.warning("No permissions found in the manifest for $permission")
            call.resolve(JSObject().put("result", PermissionStatus.DENIED.rawValue))
            return
        }

        shouldShowRationale = ActivityCompat.shouldShowRequestPermissionRationale(activity, manifestPermissions.first())
        hasOnGoingPermissionRequest = true
        permissionRequestCall = call

        permissionsLauncher.launch(manifestPermissions.toTypedArray())
    }

    @PluginMethod
    public fun openAppSettings(call: PluginCall) {
        try {
            val packageName = Uri.fromParts("package", context.packageName, null)
            context.startActivity(
                Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, packageName)
                    .addCategory(Intent.CATEGORY_DEFAULT)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    .addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY)
                    .addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS)
            )

            call.resolve()
        } catch (e: Exception) {
            call.reject("Unable to open the app settings.", e)
        }
    }

    // region NotificareGeo.Listener

    override fun onLocationUpdated(location: NotificareLocation) {
        try {
            EventBroker.dispatchEvent("location_updated", location.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the location_updated event.", e)
        }
    }

    override fun onRegionEntered(region: NotificareRegion) {
        try {
            EventBroker.dispatchEvent("region_entered", region.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the region_entered event.", e)
        }
    }

    override fun onRegionExited(region: NotificareRegion) {
        try {
            EventBroker.dispatchEvent("region_exited", region.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the region_exited event.", e)
        }
    }

    override fun onBeaconEntered(beacon: NotificareBeacon) {
        try {
            EventBroker.dispatchEvent("beacon_entered", beacon.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the beacon_entered event.", e)
        }
    }

    override fun onBeaconExited(beacon: NotificareBeacon) {
        try {
            EventBroker.dispatchEvent("beacon_exited", beacon.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the beacon_exited event.", e)
        }
    }

    override fun onBeaconsRanged(region: NotificareRegion, beacons: List<NotificareBeacon>) {
        try {
            val data = JSObject()
            data.put("region", region.toJson())
            data.put("beacons", JSArray().apply {
                beacons.forEach { put(it.toJson()) }
            })

            EventBroker.dispatchEvent("beacons_ranged", data)
        } catch (e: Exception) {
            logger.error("Failed to emit the beacons_ranged event.", e)
        }
    }

    // endregion

    private fun determinePermissionStatus(context: Context, permission: PermissionGroup): PermissionStatus {
        if (permission == PermissionGroup.BLUETOOTH) {
            return checkBluetoothPermissionStatus(context)
        }

        if (permission == PermissionGroup.BLUETOOTH_SCAN && Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
            return checkBluetoothPermissionStatus(context)
        }

        val manifestPermissions = getManifestPermissions(context, permission)

        // If no permissions were found there's an issue and the permission is not set in the Android Manifest.
        if (manifestPermissions.isEmpty()) {
            logger.warning("No permissions found in the manifest for $permission")
            return PermissionStatus.DENIED
        }

        val granted = manifestPermissions.all {
            ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
        }

        return if (granted) PermissionStatus.GRANTED else PermissionStatus.DENIED
    }

    private fun getManifestPermissions(context: Context, permission: PermissionGroup): List<String> {
        val permissions = mutableListOf<String>()

        when (permission) {
            PermissionGroup.LOCATION_WHEN_IN_USE -> {
                if (hasPermissionInManifest(context, Manifest.permission.ACCESS_COARSE_LOCATION))
                    permissions.add(Manifest.permission.ACCESS_COARSE_LOCATION)

                if (hasPermissionInManifest(context, Manifest.permission.ACCESS_FINE_LOCATION))
                    permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
            }
            PermissionGroup.LOCATION_ALWAYS -> {
                // Note that the LOCATION_ALWAYS will deliberately request LOCATION_WHEN_IN_USE
                // case on pre Android Q devices. The ACCESS_BACKGROUND_LOCATION permission was only
                // introduced in Android Q, before it should be treated as the ACCESS_COARSE_LOCATION or
                // ACCESS_FINE_LOCATION.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    if (hasPermissionInManifest(context, Manifest.permission.ACCESS_BACKGROUND_LOCATION))
                        permissions.add(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
                } else {
                    permissions.addAll(getManifestPermissions(context, PermissionGroup.LOCATION_WHEN_IN_USE))
                }
            }
            PermissionGroup.BLUETOOTH_SCAN -> {
                // The BLUETOOTH_SCAN permission is introduced in Android S, meaning we should
                // not handle permissions on pre Android S devices.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    if (hasPermissionInManifest(context, Manifest.permission.BLUETOOTH_SCAN))
                        permissions.add(Manifest.permission.BLUETOOTH_SCAN)
                } else {
                    permissions.addAll(getManifestPermissions(context, PermissionGroup.BLUETOOTH))
                }
            }
            PermissionGroup.BLUETOOTH -> {
                if (hasPermissionInManifest(context, Manifest.permission.BLUETOOTH))
                    permissions.add(Manifest.permission.BLUETOOTH)
            }
        }

        return permissions
    }

    private fun hasPermissionInManifest(context: Context, permission: String): Boolean {
        try {
            val info = context.packageManager.getPackageInfo(context.packageName, PackageManager.GET_PERMISSIONS)
            if (info == null) {
                Log.d(TAG, "Unable to get Package info, will not be able to determine permissions to request.")
                return false
            }

            for (r in info.requestedPermissions) {
                if (r == permission) {
                    return true
                }
            }
        } catch (ex: Exception) {
            Log.d(TAG, "Unable to check manifest for permission: ", ex)
        }

        return false
    }

    private fun checkBluetoothPermissionStatus(context: Context): PermissionStatus {
        val manifestPermissions = getManifestPermissions(context, PermissionGroup.BLUETOOTH)

        if (manifestPermissions.isEmpty()) {
            logger.warning("Bluetooth permission missing in the manifest.")
            return PermissionStatus.DENIED
        }

        return PermissionStatus.GRANTED
    }

    internal enum class PermissionGroup {
        LOCATION_WHEN_IN_USE,
        LOCATION_ALWAYS,
        BLUETOOTH_SCAN,
        BLUETOOTH;

        internal val rawValue: String
            get() = when (this) {
                LOCATION_WHEN_IN_USE -> "location_when_in_use"
                LOCATION_ALWAYS -> "location_always"
                BLUETOOTH_SCAN -> "bluetooth_scan"
                BLUETOOTH -> "bluetooth"
            }

        internal companion object {
            internal fun parse(permission: String): PermissionGroup? {
                values().forEach {
                    if (it.rawValue == permission) return it
                }

                return null
            }
        }
    }

    internal enum class PermissionStatus {
        DENIED,
        GRANTED,
        PERMANENTLY_DENIED;

        internal val rawValue: String
            get() = when (this) {
                DENIED -> "denied"
                GRANTED -> "granted"
                PERMANENTLY_DENIED -> "permanently_denied"
            }

        internal companion object {
            internal fun parse(status: String): PermissionStatus? {
                values().forEach {
                    if (it.rawValue == status) return it
                }

                return null
            }
        }
    }
}
